### PR TITLE
Windows CI: Fail if busybox image fails to build

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

Couple of tweaks:

On an internal run where a server was unavailable, the busybox image failed to build, but the scripts continued regardless to attempt to run the CLI suite (which obviously failed). This is because the `set -e` was in the LATESTCOUNT if statement so a docker build failure didn't error out. Moved it outside the `if` block.

Also noticed that the file had a redundant `set -e` at the start as it was immediately followed by a `set +e`.

@tianon

### CODE REVIEW

**Comment by jhowardmsft** [comment]
No idea what's going on with experimental - looks like it passed all the CLI tests then timed out after it completed. Has been going on with multiple PRs.

**Comment by jhowardmsft** [comment]
Ping @tianon :smiley:

**Comment by tonistiigi** [comment]
SGTM

**Comment by tianon** [comment]
I'm always -1 on any use of `set +e` (put another way, the absence of `set -e`), since it being a solution usually means the script isn't handling failure appropriately (and unexpected failures usually _should_ stop the script from executing).

**Comment by tianon** [comment]
For example, that first `if` should be replaced by something like this:
```bash
if ! docker images | grep -v latest | grep -q windowsservercore; then
  ...
fi
```

**Comment by tianon** [comment]
Give me a few minutes, and I'll work up an alternative to this. :smile:

**Comment by tianon** [comment]
I _think_ I've understood the intent of this file (and reflected it below):

```bash
#!/bin/bash
set -e

if ! docker inspect windowsservercore:latest &> /dev/null; then
	latestBuild="$(docker images | awk -F '  +' 'NR > 1 && $1 ~ /windowsservercore$/ && $2 != "latest" { print $1 ":" $2; exit }')"
	if [ -z "$latestBuild" ]; then
		echo "ERROR: Could not find windowsservercore images"
		exit 1
	fi

	( set -x && docker tag "$latestBuild" windowsservercore:latest )
fi

if ! docker inspect busybox:latest &> /dev/null; then
	(
		set -x
		docker build -t busybox - <<'EOF'
			FROM windowsservercore
			RUN mkdir C:\tmp && mkdir C:\busybox
			ADD http://frippery.org/files/busybox/busybox-w32-FRP-483-g31277ab.exe /busybox/busybox.exe
			RUN setx /M PATH "C:\busybox;%PATH%"
			RUN powershell busybox.exe --list ^|%{$nul = cmd /c mklink C:\busybox\$_.exe busybox.exe}
			ENTRYPOINT ["C:/busybox/busybox.exe"]
		EOF
	)
fi
```

In layman's terms -- if `windowservercore:latest` doesn't exist, find the most recent `*windowservercore` image and tag it.

Then, if `busybox:latest` doesn't exist, build it from a simple `Dockerfile` (embedded instead of fetched so that @jhowardmsft accidentally or intentionally changing his `master` branch doesn't break anything here, and using a specific version from frippery.org instead of just hoping that the main `busybox.exe` doesn't change/break :innocent:).

**Comment by jhowardmsft** [comment]
Thanks @tianon for the suggestions. However, I'm closing this PR as moving forward, to allow us to run CI on nanoserver, I've been porting the overall CI script from bash to Powershell. That's imminently about to go live in Jenkins, and as a result, this file will no longer be used and will be removed from the docker repo.
